plugins/openstack/src/main/java/org/eclipse/osc/orchestrator/plugin/openstack/OpenstackOrchestratorPlugin.java

```diff
@@ -6,90 +6,145 @@
-            writer.write(authJson);
-            writer.flush();
+        Identifier projectIdentifier = null;
+        if (projectId != null) {
+            projectIdentifier = Identifier.byId(projectId);
         }
-        if (connection.getResponseCode() != 201) {
-            throw new IllegalStateException("Can't get token (" + connection.getResponseCode() + "): " + connection.getResponseMessage());
+        if (projectName != null) {
+            projectIdentifier = Identifier.byName(projectName);
+        }
+
+        if (domainIdentifier != null && projectIdentifier != null) {
+            osClient = OSFactory.builderV3()
+                    .endpoint(endpoint)
+                    .credentials(username, secret)
+                    .scopeToProject(projectIdentifier, domainIdentifier)
+                    .authenticate();
+        }
+        if (domainIdentifier == null && projectIdentifier != null) {
+            osClient = OSFactory.builderV3()
+                    .endpoint(endpoint)
+                    .credentials(username, secret)
+                    .scopeToProject(projectIdentifier)
+                    .authenticate();
+        }
+        if (domainIdentifier != null && projectIdentifier == null) {
+            osClient = OSFactory.builderV3()
+                    .endpoint(endpoint)
+                    .credentials(username, secret)
+                    .scopeToDomain(domainIdentifier)
+                    .authenticate();
         }
-        return connection.getHeaderField("X-Subject-Token");
     }
 
     @Override
     public void registerManagedService(Ocl ocl) {
-        log.info("Register managed service, creating openstack resource");
+        log.info("Register managed service, creating openstack resources");
+
+        log.info("Creating Neutron network resources ...");
+        ocl.getNetwork().getSubnet().forEach(subnet -> {
+            osClient.networking().subnet().create(Builders.subnet()
+                    .name(subnet.getId())
+                    .ipVersion(IPVersionType.V4)
+                    .cidr(subnet.getCidr())
+                    .build());
+        });
+        ocl.getImage().getArtifacts().forEach(artifact -> {
+            if (artifact.getType().equalsIgnoreCase("docker")) {
+                log.info("Starting docker container via Magnum ...");
+                ContainerBuilder builder = new MagnumContainer.ContainerConcreteBuilder();
+                builder.image(artifact.getBase());
+                builder.name(artifact.getName());
+                Container container = osClient.magnum().createContainer(builder.build());
+                osClient.magnum().startContainer(artifact.getName());
+                log.info("Docker container " + container.getStatus());
+            }
+            if (artifact.getType().equalsIgnoreCase("image")) {
+                log.info("Starting bare image via Glance ...");
+                Image image = osClient.imagesV2().create(Builders.imageV2()
```

Comment:
From OSC, should we consider even uploading image or the assumption would be that the images already exist in openstack and we create only a VM out of it?

Reply:
Agree to refactor, actually we have three types of images for now, `image`, `container`, and `artifact` (image+provisioners). 
For the `artifact`, how about we embed the base image into the `artifact` block directly?

Reply:
@iskey Thanks. For sure we need some rework on this area of OCL data model. Let's discuss this as part of next sprint tasks.